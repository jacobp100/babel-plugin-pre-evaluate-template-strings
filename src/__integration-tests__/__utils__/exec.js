/* @flow */

import childProcess from 'child_process';
import path from 'path';

export function transpile(
  source: string,
  pluginOptions: any = { cache: false, extract: false },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_TRANSPILE_BIN = path.join(__dirname, './transpile.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_TRANSPILE_BIN} '${JSON.stringify(
      source
    )}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [transpiled, rawStyles] = stdout.toString().split('\n');
  const styles = JSON.parse(rawStyles || '{}');
  return {
    code: JSON.parse(transpiled || '""'),
    styles,
    getCSSForClassName(className: string) {
      return styles[`.${className}`];
    },
  };
}

export function extract(
  source: string,
  pluginOptions: any = { cache: false, extract: true },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_EXTRACT_BIN = path.join(__dirname, './extract.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_EXTRACT_BIN} '${JSON.stringify(source)}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [data, filenames, transpiled] = stdout.toString().split('\n');
  return {
    data: JSON.parse(data || '""'),
    filenames: JSON.parse(filenames || '[]'),
    transpiled: JSON.parse(transpiled || '""'),
  };
}
