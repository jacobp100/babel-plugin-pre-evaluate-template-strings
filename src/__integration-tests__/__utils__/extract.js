const fs = require('fs');

const writeFileSync = fs.writeFileSync;

let buffer = '';
const filenames = [];

fs.writeFileSync = (filename, data, opts) => {
  if (/\.css$/.test(filename)) {
    filenames.push(filename);
    buffer += data;
    return;
  }
  writeFileSync(filename, data, opts);
};

const transpile = require('./transpile');

const code = transpile(
  JSON.parse(process.argv[2]),
  JSON.parse(process.argv[3])
);

console.log(JSON.stringify(buffer));
console.log(JSON.stringify(filenames));
console.log(JSON.stringify(code || ''));
