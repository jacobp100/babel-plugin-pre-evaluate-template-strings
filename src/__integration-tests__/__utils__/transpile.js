const babel = require('babel-core');
const path = require('path');
const dedent = require('dedent');

const CSS_PATH = require.resolve(path.join(process.cwd(), 'build/css'));
const SHEET_PATH = require.resolve(path.join(process.cwd(), 'build/sheet'));
const PREVAL_PLUGIN_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/preval-extract')
);
const MODULE_SYSTEM_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/lib/moduleSystem')
);

function transpile(source, options = { pluginOptions: {}, babelOptions: {} }) {
  return babel.transform(
    dedent`
      import css from '${CSS_PATH}';

      ${source}
    `,
    Object.assign(
      {
        presets: ['env', 'stage-2', 'react'],
        plugins: [[PREVAL_PLUGIN_PATH, options.pluginOptions]],
        sourceMaps: false,
      },
      options.babelOptions
    )
  ).code;
}

module.exports = transpile;

if (require.main && module.filename === require.main.filename) {
  const code = transpile(
    JSON.parse(process.argv[2]),
    JSON.parse(process.argv[3])
  );

  if (code) {
    console.log(JSON.stringify(code));

    // eslint-disable-next-line global-require
    const getCachedModule = require(MODULE_SYSTEM_PATH).getCachedModule;
    const sheet = getCachedModule(SHEET_PATH);
    const styles = sheet
      ? getCachedModule(SHEET_PATH).exports.default.styles()
      : {};
    console.log(JSON.stringify(styles || ''));
  }
}
