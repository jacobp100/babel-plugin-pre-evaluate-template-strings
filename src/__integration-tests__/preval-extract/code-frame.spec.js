/* eslint-disable no-template-curly-in-string */
/* @flow */

import dedent from 'dedent';

import { transpile } from '../__utils__/exec';

describe('preval-extract babel plugin code frames', () => {
  it('should build valid code frame for errors while evaling styles', () => {
    expect(() => {
      transpile(dedent`
      const test = () => {
        throw new Error("Some weird error");
      };

      function m() {
        test();
      }

      const header = css\`color: ${'${m()}'};\`;
      `);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      transpile(dedent`
      const utils = require("./src/babel/__integration-tests__/__fixtures__/commonjs/utils.js");

      const header = css\`color: ${'${utils.throw()}'};\`;
      `);
    }).toThrowErrorMatchingSnapshot();
  });
});
