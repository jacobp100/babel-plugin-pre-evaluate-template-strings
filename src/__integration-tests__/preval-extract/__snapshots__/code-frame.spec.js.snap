// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preval-extract babel plugin code frames should build valid code frame for errors while evaling styles 1`] = `
Command failed: node <<CWD>>/src/babel/__integration-tests__/__utils__/transpile.js '"const test = () => {\\n  throw new Error(\\"Some weird error\\");\\n};\\n\\nfunction m() {\\n  test();\\n}\\n\\nconst header = css\`color: \${m()};\`;"' '{"pluginOptions":{"cache":false,"extract":false},"babelOptions":{"filename":"test.js"}}'
<<CWD>>/node_modules/babel-core/lib/transformation/file/index.js:590
      throw err;
      ^

Error: Some weird error
  1 | import css from '<<CWD>>/build/css.js';
  2 | 
  3 | var test = () => {
> 4 | throw new Error("Some weird error");
    |      ^
  5 | }
  6 | 
  7 | function m() {
at <<CWD>>/test.js:4:6

`;

exports[`preval-extract babel plugin code frames should build valid code frame for errors while evaling styles 2`] = `
Command failed: node <<CWD>>/src/babel/__integration-tests__/__utils__/transpile.js '"const utils = require(\\"./src/babel/__integration-tests__/__fixtures__/commonjs/utils.js\\");\\n\\nconst header = css\`color: \${utils.throw()};\`;"' '{"pluginOptions":{"cache":false,"extract":false},"babelOptions":{"filename":"test.js"}}'
<<CWD>>/node_modules/babel-core/lib/transformation/file/index.js:590
      throw err;
      ^

Error: test
   6 |   add5(value) {
   7 |     return value + 5;
   8 |   },
   9 |   throw() {
> 10 |     throw new Error('test');
     |          ^
  11 |   },
  12 | };
  13 | 
at <<CWD>>/src/babel/__integration-tests__/__fixtures__/commonjs/utils.js:10:10
at <<CWD>>/test.js:5:58

`;
